ledger/lruboxes_test.go

```diff
@@ -0,0 +1,240 @@
+			_, has := baseBoxes.read(fmt.Sprintf("key%d", i))
+			if !has {
+				allBoxesLoaded = false
+				break
+			}
+		}
+		if allBoxesLoaded {
+			break
+		}
+		if time.Since(testStarted).Seconds() > 20 {
+			require.Fail(t, "failed after waiting for 20 second")
+		}
+		// not yet, keep looping.
+	}
+}
+
+type lruBoxesTestLogger struct {
+	logging.Logger
+	WarnfCallback func(string, ...interface{})
+	warnMsgCount  int
+}
+
+func (cl *lruBoxesTestLogger) Warnf(s string, args ...interface{}) {
+	cl.warnMsgCount++
+}
+
+func TestLRUBoxesPendingWritesWarning(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	var baseBoxes lruBoxes
+	pendingWritesBuffer := 50
+	pendingWritesThreshold := 40
+	log := &lruBoxesTestLogger{Logger: logging.TestingLog(t)}
+	baseBoxes.init(log, pendingWritesBuffer, pendingWritesThreshold)
+	for j := 0; j < 50; j++ {
+		for i := 0; i < j; i++ {
+			boxValue := fmt.Sprintf("box %d value", i)
+			box := persistedBoxData{
+				value: &boxValue,
+				round: basics.Round(i),
+			}
+			baseBoxes.writePending(box, fmt.Sprintf("key%d", i))
+		}
+		baseBoxes.flushPendingWrites()
+		if j >= pendingWritesThreshold {
+			// expect a warning in the log
+			require.Equal(t, 1+j-pendingWritesThreshold, log.warnMsgCount)
+		}
+	}
+}
+
+func TestLRUBoxesOmittedPendingWrites(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	var baseBoxes lruBoxes
+	pendingWritesBuffer := 50
+	pendingWritesThreshold := 40
+	log := &lruBoxesTestLogger{Logger: logging.TestingLog(t)}
+	baseBoxes.init(log, pendingWritesBuffer, pendingWritesThreshold)
+
+	for i := 0; i < pendingWritesBuffer*2; i++ {
+		boxValue := fmt.Sprintf("box %d value", i)
+		box := persistedBoxData{
+			value: &boxValue,
+			round: basics.Round(i),
+		}
+		baseBoxes.writePending(box, fmt.Sprintf("key%d", i))
+	}
+
+	baseBoxes.flushPendingWrites()
+
+	// verify that all these boxes are truly there.
+	for i := 0; i < pendingWritesBuffer; i++ {
+		box, has := baseBoxes.read(fmt.Sprintf("key%d", i))
+		require.True(t, has)
+		require.Equal(t, basics.Round(i), box.round)
+		require.Equal(t, fmt.Sprintf("box %d value", i), *(box.value))
+	}
+
+	// verify expected missing entries
+	for i := pendingWritesBuffer; i < pendingWritesBuffer*2; i++ {
+		box, has := baseBoxes.read(fmt.Sprintf("key%d", i))
+		require.False(t, has)
+		require.Equal(t, persistedBoxData{}, box)
+	}
+}
+
+func BenchmarkLRUBoxesWrite(b *testing.B) {
+	numTestBoxes := 5000
+	// there are 2500 boxes that overlap
+	fillerBoxes := generatePersistedBoxesData(0, 97500)
+	boxes := generatePersistedBoxesData(97500-numTestBoxes/2, 97500+numTestBoxes/2)
+
+	benchLruWriteBoxes(b, fillerBoxes, boxes)
+}
+
+func benchLruWriteBoxes(b *testing.B, fillerBoxes []cachedBoxData, boxes []cachedBoxData) {
+	b.ResetTimer()
+	b.StopTimer()
+	var baseBoxes lruBoxes
+	// setting up the baseBoxes with a predefined cache size
+	baseBoxes.init(logging.TestingLog(b), baseResourcesPendingAccountsBufferSize, baseResourcesPendingAccountsWarnThreshold) // todo
```

Comment:
@algoidurovic What does the TODO signify here?

I'm _guessing_ it's to replace the _account_ constants with _box_ constants.

Reply:
Sorry - replace _account_ with _resources_ in my prior note.

Reply:
@algoidurovic Resolving - Thanks for https://github.com/algorand/go-algorand/pull/4275/commits/abaf6ca575a9589e3791ebed53cddf8ccc5f62a3.